code/controllers/subsystem/SSthrowing.dm

```diff
@@ -145,51 +222,55 @@ SUBSYSTEM_DEF(throwing)
 			finalize()
 			return
 
-		AM.Move(step, get_dir(AM, step))
+		if(!AM.Move(step, get_dir(AM, step))) // we hit something during our move...
+			if(AM.throwing) // ...but finalize() wasn't called on Bump() because of a higher level definition that doesn't always call parent.
+				finalize()
+			return
 
 		dist_travelled++
 
-		if(dist_travelled > MAX_THROWING_DIST)
+		if(actual_target && !(actual_target.pass_flags_self & LETPASSTHROW) && actual_target.loc == AM.loc) // we crossed a movable with no density (e.g. a mouse or APC) we intend to hit anyway.
+			finalize(TRUE, actual_target)
+			return
+
+		if (dist_travelled > MAX_THROWING_DIST)
 			finalize()
 			return
 
 /datum/thrownthing/proc/finalize(hit = FALSE, target = null)
-	set waitfor = 0
-	SSthrowing.processing -= thrownthing
+	set waitfor = FALSE
 	//done throwing, either because it hit something or it finished moving
+	if(!thrownthing)
```

Comment:
why not check for qdeleted here

Reply:
`on_thrownthing_qdel()` will get triggered by COMSIG_PARENT_QDELETING before `finalize()` will be reached.

Reply:
Makes sense